Simple HEAD

ksdenisova opened the pull request:
**Feature**: Simple HEAD

  **Scenario**: Executing a HEAD request to /simple_get
    **Given** I make a HEAD request to "/simple_get"
    **Then** my response should have status code 200
    **And** my response should have an empty body

  **Scenario**: HEAD does not include body
    **Given** I make a HEAD request to "/head_request"
    **Then** my response should have status code 200
    **And** my response should have an empty body

* All existing test endpoints should be updated with the required headers

[Trello](https://trello.com/c/oG0m1tc5/12-simple-head)

<img width="1065" alt="image" src="https://user-images.githubusercontent.com/89826596/146954133-ff8b4e57-0d6c-46e8-9fd1-0cec49e5cabd.png">

**Refactoring:**

- Updated `Headers` to support `Content-Length` and `Content-Type` headers.
- Added `Content-Length` and `Content-Type` headers to `SimplePost` and `SimpleGetWithBody` handlers.
- Changed parsing the body in `Request.Parser`.

Aoxian commented on lib/http_server_fixture/handler/simple_get_with_body.ex:
```diff
@@ -1,9 +1,15 @@
   @impl HttpServer.Handler
   def handle(_request, builder) do
+    body = "Hello world"
+    headers = %Headers{content_length: byte_size(body), content_type: "text/plain"}
```
It should not be the responsibility of a user of the http_server to calculate their own body size.  What can you do in the http server (and not in the tests handlers) to do this?

ksdenisova replied:
I'll move it to the `Response` module. I just need to think how to add this only if response has a body.

Aoxian commented on lib/http_server_fixture/handler/simple_head.ex:
```diff
@@ -0,0 +1,8 @@
+defmodule HttpServerFixture.Handler.SimpleHead do
+  @behaviour HttpServer.Handler
+
+  @impl HttpServer.Handler
+  def handle(_request, builder) do
+    builder.new_response()
+  end
+end
```
Where is this being used? I did not think the http server spec requires an endpoint for /simple_head for the acceptance tests?

ksdenisova replied:
This is used in this acceptance test:

  **Scenario**: HEAD does not include body
    **Given** I make a HEAD request to "/head_request"
    **Then** my response should have status code 200
    **And** my response should have an empty body

This is actually interesting, we didn't have a `GET` method allowed for this endpoint. How should I support this? Should I return a `SimpleGet` handler for this endpoint?

Aoxian replied:
There is nothing about this acceptance test that says HEAD is the only option.  You should be able to add a GET route to `/head_request` and that means the HEAD request to the enpoint would return correclty.

Aoxian replied:
What I suggest is adding the following `Handler` to the fixture:

```elixir
defmodule HttpServerFixture.Handler.SimpleHead do
  @behaviour HttpServer.Handler

  @impl HttpServer.Handler
  def handle(_request, builder) do
    body = "This would return in a GET, but not in a HEAD"
    headers = %Headers{content_type: "text/plain"}

    builder.new_response()
    |>builder.add_headers(headers)
    |> builder.add_body(body)
  end
end
```

Aoxian replied:
I would expect a HEAD request to return the following:
```
HTTP/1.1 HEAD 200, OK
Content-Type: text/plain
Content-Length: 45
```

Aoxian replied:
I think that would help you flush out the functionality, but also pass the acceptance test.

ksdenisova replied:
> I would expect a HEAD request to return the following:
> 
> ```
> HTTP/1.1 HEAD 200, OK
> Content-Type: text/plain
> Content-Length: 0
> ```

I'm trying to understand if response to HEAD request should contain the representation headers. From [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages) I see that they applied if the message has a body:

`Representation headers like Content-Type that describe the original format of the message data and any encoding applied (only present if the message has a body).` 

And also that [Representation headers](https://developer.mozilla.org/en-US/docs/Glossary/Representation_header) describe the body content that would be selected if the resource was actually requested:

`Representation headers may be present in both HTTP request and response messages. If sent as a response to a HEAD request, they describe the body content that would be selected if the resource was actually requested.`

Does this mean that if the request has a body, we should include those headers in the response, and not otherwise?

Aoxian replied:
you are correct, in this case it would not have the headers.  Its a very frustrating thing that the acceptance tests are minimal.

Aoxian replied:
Here is another hypothetical, perhaps this will help understand what I am trying to communicate here. ....

```elixir
defmodule HttpServerFixture.Handler.HeadExample do
  @behaviour HttpServer.Handler

  @impl HttpServer.Handler
  def handle(_request, builder) do
    body = "This body shows up in a GET but not in a HEAD"
    headers = %Headers{content_type: "text/plain"}

    builder.new_response()
    |>builder.add_headers(headers)
    |> builder.add_body(body)
  end
end
```
Route:
```elixir
def handle(%Request{resource: "/head_example", method: method}) do
    case method do
      "GET" ->
        Handler.HeadRequest
      _ ->
        raise Exception.MethodNotFoundError
    end
  end
```

GET response:
```
200 OK
Content-Type: text/plain 
Content-Length: 45

This body shows up in a GET but not in a HEAD
```

HEAD response:
```
200 OK
```

Aoxian replied:
I really hope this is helpful feedback.  If not, we can sync up tomorrow during the weekly pairing session and work through it.

ksdenisova replied:
I understand what you mean. If I won't implement it until pairing session, then I will be glad to do it together. Thank you for the explanation and such good examples!

Aoxian commented on lib/routes.ex:
```diff
@@ -16,6 +16,8 @@ defmodule Routes do
     case method do
       "GET" ->
         Handler.SimpleGet
+      "HEAD" ->
+        Handler.SimpleHead
```
a routes file, which is essentially a configuration file for your implementation, should not have HEAD as a method. The http server library / framework you are writing should handle the HEAD request for any GET that is implemented.

ksdenisova replied:
I understand it now. I'll think about how I can approach this differently.

Aoxian reviewed (changes requested):
The approach taken in it current state follows the pattern of placing the burden of implementing a HEAD method on the user of the library / framework. Instead, when a user implements a GET route, HEAD should be provided for them by the http server library.  Very similar to how you have handled when a route is not found.  Please refactor.

ksdenisova commented:
I moved `Content-Length` calculation to `Response`, refactored `Router` to support `HEAD` request. I want to refactor `find_handler` to more functional way.

I checked that `body` is empty when making `HEAD` request to `head_request` endpoint using Postman: 

<img width="333" alt="image" src="https://user-images.githubusercontent.com/89826596/147136009-f9a8f8f9-3bd7-4ee0-9b97-df2ece75343b.png">

But when I  make `GET` request to the same endpoint, response has a `body`:

<img width="315" alt="image" src="https://user-images.githubusercontent.com/89826596/147136573-19d9e5dc-742d-4d06-a24a-8948a56b3a8e.png">

I still need to figure out how not to send Content length/type if body is empty. I'll continue to work on this tomorrow.

Aoxian commented on lib/http_server/response/response.ex:
```diff
@@ -24,4 +24,10 @@ defmodule HttpServer.Response.Response do
     }[code]
   end
+
+  def update_headers(response) when response.body != nil do
+    %{response | headers: %{response.headers | content_length: byte_size(response.body)}}
+  end
+
+  def update_headers(response), do: response
```
(this is a non-blocking comment)

When is it better to use guards and when is it better to pattern match?

For example, here is an alternate solution using pattern matching:
```suggestion
  def update_headers(%Response{body: nil} = response), do: response

  def update_headers(%Response{headers: headers, body: body} = response) do
    %Response{response | headers: %{headers | content_length: byte_size(body)}}
  end

  
```

ksdenisova replied:
We use pattern match to check if some data (variables, maps, structs etc.) matches a specific pattern.

Guards are more powerful that pattern match. They uses pattern matching to achieve more complex checks: 
> [Guards](https://hexdocs.pm/elixir/master/patterns-and-guards.html) are a way to augment pattern matching with more complex checks. They are allowed in a predefined set of constructs where pattern matching is allowed, such as function definitions, case clauses, and others.

I think in our case it is enough to use only patterns. I'll make changes.

Aoxian commented on lib/http_server/response/response.ex:
```diff
@@ -24,4 +24,10 @@ defmodule HttpServer.Response.Response do
     }[code]
   end
+
+  def update_headers(response) when response.body != nil do
```
If we are expecting a %Response{} struct to be passed in, you can make sure that the function pattern matches on it, to prevent incorrect data being passed in. (you probably want to do this anytime in the code you are passing around a custom Struct.

```suggestion
  def update_headers(%Response{} = response) when response.body != nil do
```

Then, anytime `#update_headers/1` is called with something other than a response you could ignore it, or maybe even raise an error.

```exlixir
  def update_headers(_), do: nil
```

ksdenisova replied:
This is a good suggestion. I'll change it and will keep it in mind for the future.

Aoxian commented on lib/http_server_fixture/handler/simple_post.ex:
```diff
@@ -1,9 +1,14 @@
+    headers = %Headers{content_type: "text/plain"}
+
     builder.new_response()
+    |>builder.add_headers(headers)
```
```suggestion
    |> builder.add_headers(headers)
```
![image](https://user-images.githubusercontent.com/26009042/147152513-6f02eee3-975b-4c73-879e-120e744da251.png)

Aoxian replied:
I am actually surprised the linter did not find this.

ksdenisova replied:
It's possible that this rule is disabled. I'll take a look.

ksdenisova replied:
No, the rule `Credo.Check.Consistency.SpaceAroundOperators, []` is active, but it looks like it checks space around operators like +, -, * and /. It's possible doesn't check if after the pipe operator.

Aoxian replied:
Sounds like an opportunity for an open source PR for Credo (assuming its an open source project).

Aoxian commented on lib/http_server/router.ex:
```diff
@@ -14,16 +14,34 @@ defmodule HttpServer.Router do
 
   def find_handler(routes \\ Routes, request) do
-    routes.handle(request)
+    case request.method do
+      "HEAD" ->
+        %{request | method: "GET"}
+        |> routes.handle()
+      _ ->
+        routes.handle(request)
+    end
```
Any time you have a case, its an opportunity to use pattern matching.  

```elixir
def find_handler(routes \\ Routes, %Request{method: "HEAD"} = request) do
  ...
end

def find_handler(routes \\ Routes, %Request{} = request), do: routes.handle(request)

def find_handler(_, _), do: nil
```

ksdenisova replied:
I actually tried to implement something similar yesterday, but I had an error about using default argument (`Routes`) twice. I'm going to find a way to use a pattern matching here.

Aoxian replied:
this is not blocking, so if you cannot figure something out just let us know!

ksdenisova replied:
I figured it out. I added a header with default values as Mix suggested.

<img width="1019" alt="CleanShot 2021-12-23 at 10 45 26@2x" src="https://user-images.githubusercontent.com/89826596/147279505-e97a0949-cfa6-4701-91fa-6a76fddac013.png">

Aoxian commented on lib/http_server/router.ex:
```diff
@@ -18,23 +19,24 @@ defmodule HttpServer.Router do
-      _ ->
-        routes.handle(request)
-    end
+  def find_handler(routes \\ Routes, request)
+
+  def find_handler(routes, %Request{method: "HEAD"} = request) do
+    %Request{request | method: "GET"}
+    |> routes.handle()
```
This is looking good.  Since we know this is for `HEAD`, why not just strip the headers right here and simplify things.
```suggestion
  def find_handler(routes \\ Routes, request)

  def find_handler(routes, %Request{method: "HEAD"} = request) do
    response = %Request{request | method: "GET"} |> routes.handle()
    
    %Response{ response | body: nil }
```

Aoxian replied:
I see you have this in the `Response`.  I think that actually makes more sense, it should not be the routers job to strip the body.

Aoxian reviewed (approved):
Nice effort!

ksdenisova commented:
> Nice effort!

Thank you for helping me with this!!
